### PR TITLE
[MV-428] Benchmark and Port MultilabelSoftmarginloss

### PR DESCRIPTION
**Pull request by littlecutebird**
- Link to issue: https://moreh.atlassian.net/browse/MV-428
- Port 2 kernels: MultilabelSoftMarginLossForward, MultilabelSoftMarginLossForwardUnreduced. Given input tensor is (N,C), MIOpen is better if C < 25. Backward is not better in general.
- New API is guarded by MIOPEN_BETA_API macro.
- See detailed benchmark result [here](https://docs.google.com/spreadsheets/d/18XnXP7nfqdIK_xcO7TdLcga3DvCv_QxpwOOxZT2VhEw/edit?usp=sharing)

- Compared to ROCm:

Unreduced: 
|type    |Forward| 
|--------|-------|
|float32 | 4.29| 
|float16 | 5.18  | 
|bfloat16| 5.56  |

Reduced: 
|type    | Forward |
|--------|-------|
|float32 | 6.13  | 
|float16 | 6.60 | 
|bfloat16| 6.59 |

### CODE REVIEW

File: driver/multilabelsoftmarginloss_driver.hpp [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+        size_t o_sz = GetTensorSpace(oDesc);
+        if(divisor != 0)
+        {
+            miopenGetMultilabelSoftMarginLossForwardWorkspaceSize(
+                GetHandle(),
+                iDesc,
+                tDesc,
+                wDesc,
+                oDesc,
+                divisor == 1 ? MIOPEN_LOSS_REDUCTION_SUM : MIOPEN_LOSS_REDUCTION_MEAN,
```
Comment (DuongQLee) [inline driver/multilabelsoftmarginloss_driver.hpp]: I think you should assign a private var "miopenLossReductionMode_t reductionMode", initialize the reductionMode based on the command line arg then pass the reductionMode to the function. Same for other functions call with the reduction mode

File: src/include/miopen/multilabelsoftmarginloss/invoke_params.hpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+struct InvokeParams : public miopen::InvokeParams
+{
+    InvokeParams() = default;
+
+    const TensorDescriptor* iDesc  = nullptr;
+    const TensorDescriptor* tDesc  = nullptr;
+    const TensorDescriptor* wDesc  = nullptr;
+    const TensorDescriptor* oDesc  = nullptr;
+    const TensorDescriptor* dODesc = nullptr;
+    const TensorDescriptor* dIDesc = nullptr;
```
Comment (DuongQLee) [inline src/include/miopen/multilabelsoftmarginloss/invoke_params.hpp]: I think you missed this when deleting backward.

File: src/include/miopen/multilabelsoftmarginloss/problem_description.hpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "MultilabelSoftMarginLoss: Input tensor need to be 2D tensor");
+        }
+        if(tDesc.GetLengths() != iDesc.GetLengths())
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "MultilabelSoftMarginLoss: Tensor tensor need to be 2D tensor which is "
+                         "the same shape as input tensor");
+        }
```
Comment (DuongQLee) [inline src/include/miopen/multilabelsoftmarginloss/problem_description.hpp:66]: I think you need to change the methods of comparing two descriptor lenghts

File: test/gtest/multilabelsoftmarginloss.hpp [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        divisor = input.desc.GetLengths()[0];
+
+        ws_sizeInBytes = miopen::GetMultilabelSoftMarginLossForwardWorkspaceSize(
+            handle,
+            input.desc,
+            target.desc,
+            weight.desc,
+            output.desc,
+            divisor == 1 ? MIOPEN_LOSS_REDUCTION_SUM : MIOPEN_LOSS_REDUCTION_MEAN);
+        if(ws_sizeInBytes == static_cast<size_t>(-1))
```
Comment (DuongQLee) [inline test/gtest/multilabelsoftmarginloss.hpp]: same as driver, I think you should make a reductionMode variable

File: include/miopen/miopen.h [hunk 5]
```diff
@@ -512,7 +512,7 @@ typedef enum
     miopenActivationABS      = 5, /*!< Absolute value \f$abs(x)\f$ */
     miopenActivationPOWER = 6, /*!< Scaled and shifted power \f$(\alpha + \beta * x)^{gamma}\f$ */
     miopenActivationCLIPPEDRELU =
-        7, /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
+        7,                     /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
```
Comment (o2buzzle) [inline include/miopen/miopen.h]: Please undo this auto-formatted change. It is not relevant to the operation being added

File: src/kernels/MIOpenMultilabelSoftMarginLoss.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        FLOAT_ACCUM t = CVT_FLOAT2ACCUM(T[T_tv.get_tensor_view_idx({n, c})]);
+
+        loss += -w * calc_loss(i, t);
+    }
+
+    O[O_tv.get_tensor_view_idx({n})] = CVT_ACCUM2FLOAT(loss / C);
+}
+
+extern "C" __global__ void
+MultilabelSoftMarginLossUnreducedForward2d(const INPUT_TYPE* __restrict__ I,
```
Comment (o2buzzle) [inline src/kernels/MIOpenMultilabelSoftMarginLoss.cpp]: If you are not going to have input and output types that differs, imo you can simply use `FLOAT` as the type for arguments. It should then be defined automatically by the compiler.

File: src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+        auto kernel        = KernelInfo{};
+        kernel.kernel_file = "MIOpenMultilabelSoftMarginLoss.cpp";
+        kernel.kernel_name = "MultilabelSoftMarginLossForward2d";
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int32_t>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int32_t>(dtype == miopenFloat)},
+            {"MIOPEN_USE_BFP16", static_cast<int32_t>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE",
```
Comment (o2buzzle) [inline src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp]: Similar to above. If you aren't expecting input and output types to differ, you can simply remove this and use `FLOAT` in the kernel

File: src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+    {
+        /* Phase 2: Reduce */
+        auto _size              = xgrid;
+        auto dtype              = problem.GetiDesc().GetType();
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int32_t>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int32_t>(dtype == miopenFloat)},
+            {"MIOPEN_USE_BFP16", static_cast<int32_t>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE",
```
Comment (o2buzzle) [inline src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp]: Same as above

**Review by o2buzzle — CHANGES_REQUESTED** [review]

File: src/kernels/MIOpenLossReduce.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "warp_shuffle.hpp"
+
+template <typename DTYPE>
+__device__ void LossSum(const DTYPE* input, DTYPE* output, size_t N)
```
Comment (o2buzzle) [inline src/kernels/MIOpenLossReduce.cpp]: These pointers should  get marked `__restrict__`, since you are not going to do pointer arithmetics with them anyway

File: src/kernels/MIOpenMultilabelSoftMarginLoss.cpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        FLOAT_ACCUM i = CVT_FLOAT2ACCUM(I[I_tv.get_tensor_view_idx({n, c})]);
+        FLOAT_ACCUM t = CVT_FLOAT2ACCUM(T[T_tv.get_tensor_view_idx({n, c})]);
+
+        loss += -w * calc_loss(i, t);
+    }
+
+    lsum[n] = CVT_ACCUM2FLOAT(loss / C / divisor);
+}
+
+extern "C" __global__ void MultilabelSoftMarginLossForward2d(const INPUT_TYPE* __restrict__ I,
```
Comment (o2buzzle) [inline src/kernels/MIOpenMultilabelSoftMarginLoss.cpp]: Same as above

**Review by o2buzzle — CHANGES_REQUESTED** [review]

File: src/kernels/MIOpenMultilabelSoftMarginLoss.cpp (hunk 6 above)
Comment (littlecutebird) [reply src/kernels/MIOpenMultilabelSoftMarginLoss.cpp]: @seungmanhan How do you think about this, should we only use FLOAT for kernel that have same type for input type and output type? I saw in the your last PR in ROCm/MIOpen https://github.com/ROCm/MIOpen/pull/2833 , in AddLayerNormForward you verify type of xDesc and yDesc need to be the same, but you still use INPUT_TYPE and OUTPUT_TYPE in AddLayernormFwdContiguous. Are there any specific reasons to do that?

File: driver/multilabelsoftmarginloss_driver.hpp (hunk 1 above)
Comment (littlecutebird) [reply driver/multilabelsoftmarginloss_driver.hpp]: fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/7d2ecff0b467e988d82ee7331ef38b58b9091a7e)

File: src/include/miopen/multilabelsoftmarginloss/invoke_params.hpp (hunk 2 above)
Comment (littlecutebird) [reply src/include/miopen/multilabelsoftmarginloss/invoke_params.hpp]: fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/7d2ecff0b467e988d82ee7331ef38b58b9091a7e)

File: test/gtest/multilabelsoftmarginloss.hpp (hunk 4 above)
Comment (littlecutebird) [reply test/gtest/multilabelsoftmarginloss.hpp]: fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/7d2ecff0b467e988d82ee7331ef38b58b9091a7e)

File: include/miopen/miopen.h (hunk 5 above)
Comment (littlecutebird) [reply include/miopen/miopen.h]: fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/7d2ecff0b467e988d82ee7331ef38b58b9091a7e)

File: src/kernels/MIOpenLossReduce.cpp (hunk 9 above)
Comment (littlecutebird) [reply src/kernels/MIOpenLossReduce.cpp]: fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/7d2ecff0b467e988d82ee7331ef38b58b9091a7e)

File: src/kernels/MIOpenMultilabelSoftMarginLoss.cpp (hunk 6 above)
Comment (seungmanhan) [reply src/kernels/MIOpenMultilabelSoftMarginLoss.cpp]: You don't have to worry about the previous work because it was done by referring to other miopen codes. In fact, the rules are something we continue to make. In fact, since most kernels in moDNN use the same input and output types, I think it is better to work according to o2buzzle's opinion.

File: test/gtest/multilabelsoftmarginloss.hpp [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+        // Tensor with 1 element to store result after reduce
+        output = tensor<T>{std::vector<size_t>{1}};
+        std::fill(output.begin(), output.end(), 0);
+
+        ref_output = tensor<T>{std::vector<size_t>{1}};
+        std::fill(ref_output.begin(), ref_output.end(), 0);
+
+        // Change this to MIOPEN_LOSS_REDUCTION_SUM to test sum reduction
+        reduction_mode = MIOPEN_LOSS_REDUCTION_MEAN;
```
Comment (seungmanhan) [inline test/gtest/multilabelsoftmarginloss.hpp]: It seems better to work on this part by receiving it from the MultilabelSoftMarginLossTestConfigs function.

**Review by seungmanhan — COMMENTED** [review]
A review of the kernel and calling the kernel has been completed. What tests did you refer to to measure performance?

**Comment by littlecutebird** [comment]
> A review of the kernel and calling the kernel has been completed. What tests did you refer to to measure performance?

There is a link to the sheet that showed detailed benchmark result. Please check the pull request description

**Comment by seungmanhan** [comment]
> > A review of the kernel and calling the kernel has been completed. What tests did you refer to to measure performance?
> 
> There is a link to the sheet that showed detailed benchmark result. Please check the pull request description

I saw that the model name was random. Isn't this the config you actually got?

**Comment by littlecutebird** [comment]
> > > A review of the kernel and calling the kernel has been completed. What tests did you refer to to measure performance?
> > 
> > 
> > There is a link to the sheet that showed detailed benchmark result. Please check the pull request description
> 
> I saw that the model name was random. Isn't this the config you actually got?

Yes, my operation doesn't have config in [Backend_AMD](https://morehio-my.sharepoint.com/:f:/g/personal/kyeonghwan_ryu_moreh_io/ErWCI9UeFgZPk3_QR6l381IBvty-BNl4wqSSTv2wwFcPJA?e=2p9nVU). I referenced some other loss operations but I found that (1) those ops didn't related to my op or (2) the config was too small. So I made random tests which contains both small, medium and big number. And actually I found that most of loss operations that we're working on didn't have configs.

And actually I have another benchmark file which I did when I started receiving this ticket. The config is from SmoothL1Loss config, which is the only operation I found somewhat similar to my operation. You can check [here](https://docs.google.com/spreadsheets/d/1AEvuVbqLL2XfvbCsFrtBbd2HVeO8S6MW/edit?usp=sharing&ouid=116880339212664542101&rtpof=true&sd=true). There are only very small configs, and the second dimension (which will be num_class in our op) is always 4.

File: test/gtest/multilabelsoftmarginloss.hpp (hunk 11 above)
Comment (littlecutebird) [reply test/gtest/multilabelsoftmarginloss.hpp]: Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/f4ae3458f5a32f213d8f94c64d822ea2696f6f8a)

File: src/kernels/MIOpenMultilabelSoftMarginLoss.cpp (hunk 10 above)
Comment (littlecutebird) [reply src/kernels/MIOpenMultilabelSoftMarginLoss.cpp]: Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)

File: src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp (hunk 8 above)
Comment (littlecutebird) [reply src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp]: Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)

File: src/kernels/MIOpenMultilabelSoftMarginLoss.cpp (hunk 6 above)
Comment (littlecutebird) [reply src/kernels/MIOpenMultilabelSoftMarginLoss.cpp]: Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)

File: src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp (hunk 7 above)
Comment (littlecutebird) [reply src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp]: Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)

**Review by seungmanhan — COMMENTED** [review]
LGTM

**Comment by DuongQLee** [comment]
LGTM!

**Review by o2buzzle — APPROVED** [review]
Looks good to me
